atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/CassandraImitatingConsensusForgettingStore.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.pue;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.palantir.atlasdb.futures.AtlasFutures;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.CheckAndSetException;
+import com.palantir.atlasdb.keyvalue.api.KeyAlreadyExistsException;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import javax.annotation.concurrent.NotThreadSafe;
+import org.immutables.value.Value;
+
+@SuppressWarnings("all")
+@NotThreadSafe
+public class CassandraImitatingConsensusForgettingStore implements ConsensusForgettingStore {
+    private final double probabilityOfFailure;
+    private final ThreadLocalRandom random = ThreadLocalRandom.current();
+    private final int quorum = 3;
+    private final AtomicLong timestamps = new AtomicLong(0);
+    private final List<Map<Cell, BytesAndTimestamp>> nodes = IntStream.range(0, 5)
+            .mapToObj(_ignore -> new HashMap<Cell, BytesAndTimestamp>())
+            .collect(Collectors.toList());
+
+    public CassandraImitatingConsensusForgettingStore(double probabilityOfFailure) {
+        this.probabilityOfFailure = probabilityOfFailure;
+    }
+
+    @Override
+    public void putUnlessExists(Cell cell, byte[] value) throws KeyAlreadyExistsException {
+        shuffleNodes();
```

Comment:
In general I'd avoid mutability where feasible. I'd probably implement this with a method that returns an Iterable of Integer that selects the participating nodes. Or (see later) if we replace the maps with an object modelling an individual node, it should just select the nodes that we want.

Then we're pretty close to making this thread-safe: I'm almost tempted to say we want an object wrapping the `Map<Cell, BytesAndTimestamp>` that's backed by a concurrent map plus _maybe_ does some synchronization if needed.